libtransmission/copy-test.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * This file is copyright Dr Robert Harvey Crowston, 2019.
+ * <crowston@protonmail.com>
+ *
+ * This file may be used and distributed under the GNU GPL versions 2 or 3 or
+ * any future version of the GPL.
+ *
+ */
+
+#include <string.h> /* memcmp() */
+
+#include "transmission.h"
+#include "crypto-utils.h"
+#include "file.h"
+#include "tr-assert.h"
+
+#include "libtransmission-test.h"
+
+/***
+****
+***/
+
+#define RANDOM_FILE_LENGTH  (1024 * 1024 * 128) /* 128 MiB */
+
+char* sandbox_path;
+char* reference_file;
+
+/***
+****
+***/
+
+static uint64_t fill_buffer_from_fd(tr_sys_file_t fd, uint64_t bytes_remaining, char* buf, size_t buf_len)
+{
+    memset(buf, 0, buf_len);
+
+    size_t buf_pos = 0;
+    while (buf_pos < buf_len && bytes_remaining > 0)
+    {
+        uint64_t const chunk_size = MIN(buf_len - buf_pos, bytes_remaining);
+        uint64_t bytes_read = 0;
+
+        tr_sys_file_read(fd, buf + buf_pos, chunk_size, &bytes_read, NULL);
+
+        TR_ASSERT(buf_pos + bytes_read <= buf_len);
+        TR_ASSERT(bytes_read <= bytes_remaining);
+        buf_pos += bytes_read;
+        bytes_remaining -= bytes_read;
+    }
+
+    return bytes_remaining;
+}
+
+static int files_are_identical(char const* fn1, char const* fn2)
+{
+    tr_sys_file_t fd1 = tr_sys_file_open(fn1, TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0, NULL);
+    tr_sys_file_t fd2 = tr_sys_file_open(fn2, TR_SYS_FILE_READ | TR_SYS_FILE_SEQUENTIAL, 0, NULL);
+    check(fd1 != TR_BAD_SYS_FILE && fd2 != TR_BAD_SYS_FILE);
```

Comment:
Don't combine conditions with `&&` when asserting, it's better to split them in two checks so that it's clear right away which one has failed (if/when it does).